Readme: add troubleshooting section (#113)
leikoilja/ha-google-home: README.md

## Troubleshooting

### Collecting useful log data

Here are the steps to generate useful log data:

1. Temporary log level change.
   1. Visit [Home Assistant Developer Services Tool](https://my.home-assistant.io/redirect/developer_services/).
   2. Choose `Logger: Set level` from the **Service** menu.
   3. Under **Service data** paste the following:
      ```yaml
      custom_components.google_home: debug
      glocaltokens: debug
      ```
   4. Click **Call Service**.
2. Read the log information.
   1. Visit [Home Assistant Logs](https://my.home-assistant.io/redirect/logs/).
   2. Click **Load Full Home Assistant Log**.
   3. Look for all `google_home` and `glocaltokens` entries.
3. Requesting help with the log information.
   1. Copy the log entries.
   2. Paste them into a discussion forum or bug report. Make sure to use quotation block.

### "Username/Password is incorrect"

If you get this error:

1. First verify you are using the correct Username and Password combination for that Google account.
2. Have you enabled 2 Factor Authentication on that Google account? _If so read the [2 Factor Authentication](#2-factor-authentication--app-passwords) section to continue._

> 3. We have seen some other custom components break the dependencies causing `ha-google-home` to fail authentication process. For more information please see [this issue](https://github.com/leikoilja/ha-google-home/issues/95).

Reviewer: It's just `google_home` now :)
```suggestion
3. We have seen some other custom components break the dependencies causing `google_home` to fail authentication process. For more information please see [this issue](https://github.com/leikoilja/ha-google-home/issues/95).
```

We will never collect any user sensitive data, only integration related
warnings and errors. See related [developers discussion](https://github.com/leikoilja/ha-google-home/discussions/116).

## Troubleshooting

### Collecting useful log data

Here are the steps to generate useful log data:

1. Temporary log level change.
   1. Visit [Home Assistant Developer Services Tool](https://my.home-assistant.io/redirect/developer_services/).
   2. Choose `Logger: Set level` from the **Service** menu.
   3. Under **Service data** paste the following:
      ```yaml
      custom_components.google_home: debug
      glocaltokens: debug
      ```
   4. Click **Call Service**.
2. Read the log information.
   1. Visit [Home Assistant Logs](https://my.home-assistant.io/redirect/logs/).
   2. Click **Load Full Home Assistant Log**.
   3. Look for all `google_home` entries.
   4. Also look for all `glocaltokens` entries.
3. Requesting help with the log information.
   1. Copy the log entries.
   2. Paste them into a discussion forum or bug report between the following two lines:
      > \```
      >

>       > \```

Reviewer: Also do we really want to go in such details?

Reviewer 2: I think it should be short and on point, so it is easier to understand even if you are not tech savvy 😊

Reviewer: @DurgNomis-drol is it yes or no? 🙂

Reviewer 2: Sorry, no we should not go into details 😊

Reviewer 3: Gents, @DurgNomis-drol, @KapJI some of your requests have been fulfilled in the latest push 😅

> Advanced tip for Windows users:
>
> Install [LogExpert](https://github.com/zarunbal/LogExpert/) to make finding log entries quicker.
>
> 1. Using LogExpert: Open `home-assistant.log` in your config folder.
> 2. Use `glocaltokens|google_home` as your _Text filter_ and check `Regex` to include logs from both _glocaltokens_ and _google_home_.

> 

Reviewer: Honestly, I'd remove this. This is too advanced to this section.

Reviewer 3: Honestly, me too 😅  It's too "windows" to meeee.. Sorry for that 😅

Reviewer 2: Agree 😊

### "Username/Password is incorrect"

If you get this error:

1. First verify you are using the correct Username and Password combination for that Google account.
2. Have you enabled 2 Factor Authentication on that Google account? _If so read the [2 Factor Authentication](#2-factor-authentication--app-passwords) section to continue._

> 3. After ruling out #1 and #2: _Read the [Collecting useful log data](#collecting-useful-log-data) section._

Reviewer: Let's add something about trying to disable other custom components which may interfere with auth process.

- [Google Home timers card](https://github.com/DurgNomis-drol/google_home_timers_card) by [@DurgNomis-drol](https://github.com/DurgNomis-drol)

## Troubleshooting

### Collecting useful log data

Here are the steps to generate useful log data:

1. Temporary log level change.

>     1. Visit [![Open your Home Assistant instance and show your service developer tools.](https://my.home-assistant.io/badges/developer_services.svg)](https://my.home-assistant.io/redirect/developer_services/)

Reviewer: Can you replace `my.home-assistant.io` badges with regular links please? It should be cleaner this way.

Reviewer 2: Why? I think it would be very HA to use the new MY feature 😄

Reviewer: Yeah, only badges. I think they're standing out too much.

Reviewer 3: `MY` feature is actually pretty neat. Not sure if @KapJI has tried it before, but we already use it in `Integration Setup` on README

Reviewer 2: @KapJI Yeah, they are quite big 😆 Lets hope they introduce some smaller ones in the future.

Author: > 
> 
> Can you replace `my.home-assistant.io` badges with regular links please? It should be cleaner this way.
I agree. I was trying to honor the way Home Assistant proposed it be used, but I don't like it. I will look briefly for smaller graphics. If not I will replace it with text.

    2. Choose `Logger: Set level` from the **Service** menu.
    3. Under **Service data** paste the following:
        ```yaml
        custom_components.google_home: debug
        glocaltokens: debug
        ```
    4. Click **Call Service**.
2. Read the log information.
    1. Visit [![Open your Home Assistant instance and show your Home Assistant logs.](https://my.home-assistant.io/badges/logs.svg)](https://my.home-assistant.io/redirect/logs/)
    2. Click **Load Full Home Assistant Log**.
    3. Look for all `google_home` entries.
    4. Also look for all `glocaltokens` entries.
3. Requesting help with the log information.
    1. Copy the log entries.
    2. Paste them into a discussion forum or bug report between the following two lines:
        >\```
        >
        >\```

> Advanced tips:

> > * Adding ` ```python ` to the beginning of your code block instead of just ` ``` ` adds color coding to make the logs easier for developers to read.

Author: Its personal preference. I like the color variation to help my eyes quickly jump to the info I want. But I will remove it.

> > * Use [LogExpert](https://github.com/zarunbal/LogExpert/) to make finding log entries quicker.

Author: Will do.

>   1. Open home-assistant.log in your config folder.
>   2. Use `glocaltokens|google_home` as your _Text filter_ to include logs from both _glocaltokens_ and _google_home_.

### "Username/Password is incorrect"
If you get this error:
1. First verify you are using the correct Username and Password combination for that Google account.
2. Have you enabled 2 Factor Authentication on thant Google account? _If so read the [2 Factor Authentication](#2-factor-authentication--app-passwords) section to continue._
3. After ruling out #1 and #2: _Read the [Collecting useful log data](#collecting-useful-log-data) section._

### 2 Factor Authentication / App Passwords
#### Error: "The setting you are looking for is not available for your account." at https://myaccount.google.com/apppasswords. 
This error will occur if you do not have 2 Factor Authentication (2FA) setup on your Google account.

  Here are the steps to correct this issue:
  1. If you have multiple Google accounts go to [Google.com](https://google.com) and log out of every accout!

>   (_This causes a lot of confusion! You may think you have enabled 2FA only to be enabling it on an alternate Google account!_)

Author: From personal experience I could not resolve my 2FA issues without doing so. I am always logged into 2 Google accounts: one @gmail.com and one @hanifin.net. I was convinced that I had enabled 2FA for my @gmail.com account, I tried several times in fact. But every time I got that "not available for your account" error when I went to create an app password.

>   (_This causes a lot of confusion! You may think you have enabled 2FA only to be enabling it on an alternate Google account!_)

Reviewer: When you click you photo on the top right at https://myaccount.google.com, you should see which account you're currently managing. You probably need to select the correct account on that menu but not to logout from them.

>   2. Go to [Google.com](https://google.com) and **only** log into the Google Home account you want to control with the Google Home Integration.

Reviewer: Btw, I registered a separate google account just for this and shared access to Home with it. This way you can avoid providing sensitive access to the main account to this integration. Probably you can add a not about this.

Author: I have a Google Apps for my Domain account, which isn't compatible with all of Google's services (Youtube Premium for example), so I have to also keep a @gmail.com account for use with most Google services as well.

> This error will occur if you do not have 2 Factor Authentication (2FA) setup on your Google account.
> 
>   Here are the steps to correct this issue:
>   1. (Optional) If you have multiple Google accounts it _might_ help to _temporarily_ and log out of every [Google account](https://google.com)!
>   (_This causes a lot of confusion! You may think you have enabled 2FA only to be enabling it on an alternate Google account!_)
>   2. Go to [Google.com](https://google.com) and log into the Google Home account you want to control with the Google Home Integration.
>   3. Visit [myaccount.google.com/apppasswords](https://myaccount.google.com/apppasswords). (At this point if you get the above mentioned error, please preform the _Optional_ step #1.)
>       1. Click **Select app** and enter a descriptive name such as _Google Home Integration for Home Assistant_.
>       2. Click the **Generate** button.
>       3. Copy the password and return to the Google Home Configuration screen.

Reviewer: ```suggestion
This error will occur if you do not have 2 Factor Authentication (2FA) enabled on your Google account.

  Here are the steps to resolve this issue:
  1. Open [Google Account settings](https://myaccount.google.com/).
  2. On the top right corner click the profile photo and select the account which you would like to use.
  3. Go to [Security](https://myaccount.google.com/security) page and make sure you have **2-Step Verification** turned on in **Signing in to Google** section.
  4. Then visit [App passwords](https://myaccount.google.com/apppasswords).
      1. Click **Select app** and enter a descriptive name such as _Google Home Integration for Home Assistant_.
      2. Click the **Generate** button.
      3. Copy the password and return to the Google Home Configuration screen.
```

  4. Return to Integrations at [my.home-assistant.io/redirect/integrations](https://my.home-assistant.io/redirect/integrations/).
      1. Click **Configure** on the Google Home integration.
      2. Enter your Google account username.

>       3. Paste the password into the "_Google account app password_" field.

Reviewer: Italic looks bad before `"`.
```suggestion
      3. Paste the password into the **Google account app password** field.
```

>       2. Enter your Google account username (including @gmail.com).

Reviewer: As we found this part is actually not necessary.
```suggestion
      2. Enter your Google account username.
```

Here are the steps to generate useful log data:

1. Temporary log level change.

>     1. Visit Home Assistant Developer Services Tool at [my.home-assistant.io/redirect/developer_services](https://my.home-assistant.io/redirect/developer_services/).

Reviewer: Let's make it a little nicer.
```suggestion
    1. Visit [Home Assistant Developer Services Tool](https://my.home-assistant.io/redirect/developer_services/).
```

        ```
    4. Click **Call Service**.
2. Read the log information.

>     1. Visit Home Assistant Logs at [my.home-assistant.io/redirect/logs](https://my.home-assistant.io/redirect/logs/).

Reviewer: ```suggestion
    1. Visit [Home Assistant Logs](https://my.home-assistant.io/redirect/logs/).
```

        >
        >\```

> > Advanced Windows user tip:

Reviewer: ```suggestion
> Advanced tip for Windows users:
```

>
> Install [LogExpert](https://github.com/zarunbal/LogExpert/) to make finding log entries quicker.

> >   1. Using LogExpert: Open home-assistant.log in your config folder.

Reviewer: ```suggestion
>   1. Using LogExpert: Open `home-assistant.log` in your config folder.
```

### 2 Factor Authentication / App Passwords
#### Error: "The setting you are looking for is not available for your account." at [myaccount.google.com/apppasswords](https://myaccount.google.com/apppasswords). 
This error will occur if you do not have 2 Factor Authentication (2FA) setup on your Google account.

  Here are the steps to correct this issue:
  1. (Optional) If you have multiple Google accounts it _might_ help to _temporarily_ and log out of every [Google account](https://google.com)!
  (_This causes a lot of confusion! You may think you have enabled 2FA only to be enabling it on an alternate Google account!_)
  2. Go to [Google.com](https://google.com) and log into the Google Home account you want to control with the Google Home Integration.
  3. Visit [myaccount.google.com/apppasswords](https://myaccount.google.com/apppasswords). (At this point if you get the above mentioned error, please preform the _Optional_ step #1.)
      1. Click **Select app** and enter a descriptive name such as _Google Home Integration for Home Assistant_.
      2. Click the **Generate** button.
      3. Copy the password and return to the Google Home Configuration screen.

>   4. Return to Integrations at [my.home-assistant.io/redirect/integrations](https://my.home-assistant.io/redirect/integrations/).

Reviewer: ```suggestion
  4. Return to [Integrations](https://my.home-assistant.io/redirect/integrations/).
```
